File: Project3_Python/World.py
import Organism
from Organism import OrganismsEnum, Point
import random
from Animals import Human


class World:
    PERCENT_OF_THE_WORLD = 0.1

    def __init__(self, size_of_the_world):
        self._list_of_organisms = []
        self._commentator = []
        self._a_big_warrior = None
        self._map_of_organisms = [[None] * size_of_the_world.y for _ in range(size_of_the_world.x)]
        self.__number_of_organisms = 0
        self.__size_of_world = size_of_the_world
        self.__turn = 1
        self.__number_of_animals = 0
        self.__initialize_organisms()
        # print(len(self.list_of_organisms))
        # self.print_map()

    def __is_human_alive(self):
        return self._a_big_warrior is not None

    def special_ability_on(self):
        if self._a_big_warrior:
            return self._a_big_warrior.special_ability()
        return "Human died."

    def get_all_organisms_by_type(self, organism_type = OrganismsEnum.Hogweed):
        t_organism_list = []
        for i in range(len(self._list_of_organisms)):
            if self._list_of_organisms[i].get_type_of_organism() == organism_type:
                t_organism_list.append(self._list_of_organisms[i])
        return t_organism_list

    @staticmethod
    def __sort_organisms_by_initiative(organisms):
        organisms.sort(key=lambda o: o.get_initiative(), reverse=True)

    def clear_world(self):
        for i in range(self.__size_of_world.x):
            for j in range(self.__size_of_world.y):
                self._map_of_organisms[i][j] = None

    def delete_world(self):
        self._commentator.clear()
        self._list_of_organisms.clear()
        self.clear_world()
        self.__number_of_organisms = 0
        self._a_big_warrior = None
        self.__turn = 1

    def __initialize_all_organisms(self):
        self.clear_world()
        if self._list_of_organisms:
            for organism in self._list_of_organisms:
                org_position = organism.get_position()
                self._map_of_organisms[org_position.x][org_position.y] = organism
        if self._a_big_warrior:
            warrior_position = self._a_big_warrior.get_position()
            self._map_of_organisms[warrior_position.x][warrior_position.y] = self._a_big_warrior

    def add_comment(self, message):
        self._commentator.append(message)

    def print_comments(self):
        if not self._commentator:
            print("Nothing happened.")
        else:
            for comment in self._commentator:
                print(comment)

    def make_save_file_contents(self):
        save_file_list = []
        if self._a_big_warrior:
            save_file_list.append(self._a_big_warrior.save_organism())
        if self._list_of_organisms:
            for organism in self._list_of_organisms:
                save_file_list.append(organism.save_organism())
        return save_file_list

    def _clear_comments(self):
        self._commentator.clear()

    def get_commentator(self):
        return self._commentator

    def make_turn(self, human_move):
        self._clear_comments()
        print("Turn:", self.__turn)
        self.__sort_organisms_by_initiative(self._list_of_organisms)
        human_flag = True
        self.__number_of_organisms = len(self._list_of_organisms)
        for i in range(self.__number_of_organisms):
            if i >= self.__number_of_organisms:
                break
            organism = self._list_of_organisms[i]
            if self._a_big_warrior:
                if organism.get_initiative() < self._a_big_warrior.get_initiative() and human_flag:
                    self._a_big_warrior.change_move(human_move)
                    self._a_big_warrior.action()
                    human_flag = False
            organism.action()
        if self._a_big_warrior:
            self._a_big_warrior.handle_special_ability()
        self.__turn += 1
        self.__initialize_all_organisms()
        # self.print_map()

    def remove_organism_from_list(self, organism_to_remove):
        self._list_of_organisms.remove(organism_to_remove)
        self.__number_of_organisms -= 1

    def get_size_of_world(self):
        return self.__size_of_world

    def get_organism_in_position(self, the_position):
        return self._map_of_organisms[the_position.x][the_position.y]

    def is_equal_in_the_position(self, the_organism, the_position):
        if the_position.x == -1:
            return False
        if self._map_of_organisms[the_position.x][the_position.y] is None:
            return False
        return self._map_of_organisms[the_position.x][the_position.y] == the_organism

    def null_the_position(self, the_position):
        self._map_of_organisms[the_position.x][the_position.y] = None

    def put_established_organism_to_world(self, the_organism):
        the_position = the_organism.get_position()
        self._map_of_organisms[the_position.x][the_position.y] = the_organism

    def check_if_not_occupied(self, the_position):
        return self._map_of_organisms[the_position.x][the_position.y] is None

    @staticmethod
    def _get_random_organism():
        return OrganismsEnum.values()[random.randint(0, len(OrganismsEnum.values()) - 1)]

    def _get_random_position(self):
        value_x = random.randint(0, self.__size_of_world.x - 1)
        value_y = random.randint(0, self.__size_of_world.y - 1)
        return Point(value_x, value_y)

    def __add_one_instance_of_all(self):
        for organism_enum in list(OrganismsEnum):
            self.add_organism(organism_enum)

    def add_organism(self, organism_type=None, position=None, strength=None):
        if position is None:
            position = self._get_random_position()
            try_ten_times = 10
            i = 0
            while self._map_of_organisms[position.x][position.y] is not None:
                if i == try_ten_times:
                    return
                position = self._get_random_position()
                i += 1
        if strength is None:
            new_organism = Organism.Organism.get_organism(organism_type, self, position)
        else:
            new_organism = Organism.Organism.get_organism(organism_type, self, position, strength)
        if new_organism is not None:
            self._list_of_organisms.append(new_organism)
            self._map_of_organisms[position.x][position.y] = new_organism

    def delete_human(self):
        self._a_big_warrior = None

    def add_human(self, position=None, strength=5, def_strength=5, is_ability_on=False, n_until_activation=0):
        if position is None:
            position = self._get_random_position()
            self._a_big_warrior = Human(self, position)
        else:
            self._a_big_warrior = Human(self, position, strength, def_strength, is_ability_on, n_until_activation)
        if self._a_big_warrior is not None:
            self._map_of_organisms[position.x][position.y] = self._a_big_warrior

    def __initialize_organisms(self):
        max_size = self.__size_of_world.x * self.__size_of_world.y - 1
        self.__number_of_animals = int(max_size * self.PERCENT_OF_THE_WORLD)
        self.add_human()
        self.__add_one_instance_of_all()
        for i in range(self.__number_of_animals):
            self.add_organism()
        #print(self.__number_of_animals)

    def print_map(self):
        for i in range(self.__size_of_world.x):
            for j in range(self.__size_of_world.y):
                print(self._map_of_organisms[i][j], end=' ')
            print()

File: Project2_Java/src/pl/pg/oop/project2/organisms/World.java
package pl.pg.oop.project2.organisms;

import pl.pg.oop.project2.organisms.animals.*;

import java.awt.*;

import java.util.*;
import java.util.List;

/**
 * World in which the Organisms live.
 */
public class World implements NamesOfOrganisms{
    private final List<Organism> ListOfOrganisms;

    private List<String> Commentator;
    private Human ABigWarrior;
    private final Organism[][] MapOfOrganisms;
    private int NumberOfOrganisms;
    private final Point SizeOfWorld;
    private final static double PERCENT_OF_THE_WORLD = 0.1;

    private int Turn;

    /**
     * Initializes the world with ListOfOrganisms and MapOfOrganisms.
     * @param SizeOfTheWorld Size of The world nxm.
     */
    public World(Point SizeOfTheWorld) {
        this.ListOfOrganisms = new ArrayList<>();
        this.Commentator = new ArrayList<>();
        this.SizeOfWorld = SizeOfTheWorld;
        this.MapOfOrganisms = new Organism[SizeOfWorld.x][SizeOfWorld.y];
        this.NumberOfOrganisms = 0;
        this.ABigWarrior = null;
        for(int i = 0; i < SizeOfWorld.x; i++) {
            for(int j = 0; j < SizeOfWorld.y; j++) {
                MapOfOrganisms[i][j] = null;
            }
        }
        this.Turn = 1;
        InitializeOrganisms();
    }

    /**
     * Checks if Human is Alive and well.
     * @return If Human is Alive returns true, else false.
     */
    public boolean isHumanAlive() {
        return (ABigWarrior != null);
    }

    /**
     * Turns on a special ability of Human.
     */
    public String specialAbilityOn() {
        return ABigWarrior.SpecialAbility();
    }

    /**
     * Sorts the List of Organisms by the Organisms initiative.
     * @param organisms List Of Organisms to be sorted.
     */
    private static void sortOrganismsByInitiative(List<Organism> organisms) {
        organisms.sort((o1, o2) -> {
            // Sort in descending order based on initiative
            return Integer.compare(o2.getInitiative(), o1.getInitiative());
        });
    }

    /**
     * Clears the map of world (organisms) by assigning every place with null.
     */
    private void clearWorld() {
        for(int i = 0; i < SizeOfWorld.x; i++) {
            for(int j = 0; j < SizeOfWorld.y; j++) {
                MapOfOrganisms[i][j] = null;
            }
        }
    }

    public void deleteWorld() {
        Commentator.clear();
        ListOfOrganisms.clear();
        clearWorld();
        NumberOfOrganisms = 0;
        ABigWarrior = null;
        Turn = 1;
    }

    /**
     * Initializes all Organisms that are in the ListOfOrganisms into MapOfOrganisms (places them into world).
     * Before the procedure is done also calls clearWorld() - cleaning the world.
     */
    private void initializeAllOrganism() {
        clearWorld();
        if(ListOfOrganisms != null) {
            for (Organism organism : ListOfOrganisms) {
                Point orgPosition = organism.getPosition();
                MapOfOrganisms[orgPosition.x][orgPosition.y] = organism;
            }
        }
        if(ABigWarrior != null) {
            Point warriorPosition = ABigWarrior.getPosition();
            MapOfOrganisms[warriorPosition.x][warriorPosition.y] = ABigWarrior;
        }
    }

    /**
     * Adds comment to Commentator.
     * @param message A comment to add.
     */
    public void addComment(String message) {
        Commentator.add(message);
    }

    public void printComments() {
        if(Commentator == null) {
            System.out.println("Nothing happened.");
        }
        else {
            for (String comment : Commentator) {
                System.out.println(comment);
            }
        }
    }

    public List<String> makeSaveFileContents() {
        List<String> saveFileList = new ArrayList<>();
        if(ABigWarrior != null) {
            saveFileList.add(ABigWarrior.SaveOrganism());
        }
        if(ListOfOrganisms != null) {
            for (Organism organism : ListOfOrganisms) {
                saveFileList.add(organism.SaveOrganism());
            }
        }
        return saveFileList;
    }

    private void clearComments() {
        Commentator.clear();
    }

    /**
     * @return The list of comments on what happened in the world.
     */
    public List<String> getCommentator() {
        return Commentator;
    }

    /**
     * Makes the next turn, takes the parameters that indicates in which direction will the human move.
     * @param HumanMove Next move of the human - right, left, up, down or no move.
     */
    public void makeTurn(move HumanMove) {
        clearComments();
        System.out.println("Turn: " + Turn);
        sortOrganismsByInitiative(ListOfOrganisms);
        boolean HumanFlag = true;
        NumberOfOrganisms = ListOfOrganisms.size();
        for (int i = 0; i < NumberOfOrganisms; i++) {
            Organism organism = ListOfOrganisms.get(i);
            if(ABigWarrior != null) {
                if (organism.getInitiative() < ABigWarrior.getInitiative() && HumanFlag) {
                    ABigWarrior.ChangeMove(HumanMove);
                    ABigWarrior.action();
                    HumanFlag = false;
                }
            }
            organism.action();
        }
        if(ABigWarrior != null) ABigWarrior.handleSpecialAbility();
        Turn++;
        initializeAllOrganism();
    }

    /**
     * Removes the provided Organism from ListOfOrganisms.
     * ! Does not remove the Organism from the world, needs to be done by nullThePosition().
     * @param OrganismToRemove Organism that needs to be removed.
     */
    public void RemoveOrganismFromList(Organism OrganismToRemove) {
        ListOfOrganisms.remove(OrganismToRemove);
        NumberOfOrganisms--;
    }

    /**
     * @return the size of world nxm as Point
     */
    public Point getSizeOfWorld() {
        return SizeOfWorld;
    }

    /**
     * Gets the Organism at position ThePosition.
     * @param ThePosition The position of Organism.
     * @return The Organism at provided position in MapOfOrganisms (world).
     */
    public Organism getOrganismInPosition(Point ThePosition) {
        return MapOfOrganisms[ThePosition.x][ThePosition.y];
    }

    /**
     * Checks if in the provided position is the provided Organism.
     * @param TheOrganism The Organism that we are searching for.
     * @param ThePosition The position that we are checking.
     * @return True if the Organism is in the position, else false.
     */
    public boolean IsEqualInThePosition(Organism TheOrganism, Point ThePosition) {
        if(ThePosition.x == -1) return false;
        if(MapOfOrganisms[ThePosition.x][ThePosition.y] == null) return false;
        return MapOfOrganisms[ThePosition.x][ThePosition.y].equals(TheOrganism);
    }

    /**
     * Delete the Organism from the world. Null the provided position in MapOfOrganisms.
     * @param ThePosition The position that we want to null.
     */
    public void nullThePosition(Point ThePosition) {
        MapOfOrganisms[ThePosition.x][ThePosition.y] = null;
    }

    /**
     * Put already existing Organism into the world into the specified position (position is specified as one of the variables of TheOrganism).
     * @param TheOrganism Organism that we want to put into the MapOfOrganisms.
     */
    public void putEstablishedOrganismToWorld(Organism TheOrganism) {
        Point ThePosition = TheOrganism.getPosition();
        MapOfOrganisms[ThePosition.x][ThePosition.y] = TheOrganism;
    }

    /**
     * Check if the position is not occupied in the MapOfOrganisms (world).
     * @param ThePosition The position that we want to check.
     * @return True if not occupied, false if occupied.
     */
    public boolean checkIfNotOccupied(Point ThePosition) {
        return MapOfOrganisms[ThePosition.x][ThePosition.y] == null;
    }

    /**
     * Provides you with a random Organism from OrganismsEnum.
     * @return Random OrganismsEnum.
     */
    private static OrganismsEnum getRandomOrganism() {
        Random random = new Random();
        return OrganismsEnum.values()[random.nextInt(OrganismsEnum.values().length)];
    }

    /**
     * Provide a random position within world limitations.
     * @return Random Position.
     */
    private Point getRandomPosition() {
        Random random = new Random();
        int valueX = random.nextInt(SizeOfWorld.x);
        int valueY = random.nextInt(SizeOfWorld.y);
        return new Point(valueX, valueY);
    }

    /**
     * Adds ones instance of all Organisms (except Human).
     */
    private void AddOneInstanceOfAll() {
        for(int i = 0; i < NumberOfOrg; i++) {
            AddOrganism(OrganismsEnum.values()[i]);
        }
    }

    /**
     * Add random Organism in the random position to the world.
     */
    public void AddOrganism() {
        Point position = getRandomPosition();
        int TryTenTimes = 10, i = 0;
        while(MapOfOrganisms[position.x][position.y] != null) {
            if(i == TryTenTimes) return;
            position = getRandomPosition();
            i++;
        }
        OrganismsEnum OrganismName = getRandomOrganism();
        Organism NewOrganism;
        NewOrganism = Organism.GetOrganism(OrganismName, this, position);
        if(NewOrganism != null) {
            ListOfOrganisms.add(NewOrganism);
            MapOfOrganisms[position.x][position.y] = NewOrganism;
        }
    }

    /**
     * Add Organism in the random position to the world.
     * @param OrganismName Type of Organism to add.
     */
    public void AddOrganism(OrganismsEnum OrganismName) {
        Point position = getRandomPosition();
        int TryTenTimes = 10, i = 0;
        while(MapOfOrganisms[position.x][position.y] != null) {
            if(i == TryTenTimes) return;
            position = getRandomPosition();
            i++;
        }
        Organism NewOrganism;
        NewOrganism = Organism.GetOrganism(OrganismName, this, position);
        if(NewOrganism != null) {
            ListOfOrganisms.add(NewOrganism);
            MapOfOrganisms[position.x][position.y] = NewOrganism;
        }
    }

    /**
     * Add random Organism in the position to the world.
     * @param OrganismName Type of Organism to add.
     * @param position Position at which the organism will be added.
     */
    public void AddOrganism(OrganismsEnum OrganismName, Point position) {
        Organism NewOrganism;
        NewOrganism = Organism.GetOrganism(OrganismName, this, position);
        if(NewOrganism != null) {
            ListOfOrganisms.add(NewOrganism);
            MapOfOrganisms[position.x][position.y] = NewOrganism;
        }
    }

    /**
     * Add random Organism in the random position with certain strength to the world.
     * @param OrganismName Type of Organism to add.
     * @param position Position at which the organism will be added.
     * @param strength Strength of Organism thad needs to be added.
     */
    public void AddOrganism(OrganismsEnum OrganismName, Point position, int strength) {
        Organism NewOrganism;
        NewOrganism = Organism.GetOrganism(OrganismName, this, position, strength);
        if(NewOrganism != null) {
            ListOfOrganisms.add(NewOrganism);
            MapOfOrganisms[position.x][position.y] = NewOrganism;
        }
    }

    /**
     * Delete the human, make ABigWarrior null.
     */
    public void DeleteHuman() {
       ABigWarrior = null;
    }

    /**
     * Add human to the random position to the world.
     */
    private void AddHuman() {
        Point position = getRandomPosition();
        ABigWarrior = new Human(this, position);
        MapOfOrganisms[position.x][position.y] = ABigWarrior;
    }

    /**
     * Add human to the position with certain strength to the world.
     * @param position Position.
     * @param strength The strength of Human.
     * @param DefStrength Default strength of Human.
     * @param IsAbilityOn Is Ability on.
     * @param nUntilActivation How much until possible activation.
     */
    public void AddHuman(Point position, int strength, int DefStrength, boolean IsAbilityOn, int nUntilActivation) {
        ABigWarrior = new Human(this, position, strength, DefStrength, IsAbilityOn, nUntilActivation);
        MapOfOrganisms[position.x][position.y] = ABigWarrior;
    }

    /**
     * Initialize random Organisms to the world.
     */
    private void InitializeOrganisms() {
        int maxSize = SizeOfWorld.x * SizeOfWorld.y - 1;
        int numberOfAnimals = (int)(maxSize*PERCENT_OF_THE_WORLD);
        AddHuman();
        AddOneInstanceOfAll();
        for(int i = 0; i < numberOfAnimals; i++) {
            AddOrganism();
        }
    }

    /**
     * Print the map by System.out.println().
     */
    public void printMap() {
        for(int i = 0; i < SizeOfWorld.x; i++) {
            for(int j = 0; j < SizeOfWorld.y; j++) {
                System.out.print(MapOfOrganisms[i][j] + " ");
            }
            System.out.println();
        }
    }
}
